File: IAIso-v5.0/core/iaiso-php/tests/Unit/Coordination/CoordinationTest.php
<?php

declare(strict_types=1);

namespace IAIso\Tests\Unit\Coordination;

use IAIso\Audit\MemorySink;
use IAIso\Coordination\CoordinatorException;
use IAIso\Coordination\RedisClient;
use IAIso\Coordination\RedisCoordinator;
use IAIso\Coordination\SharedPressureCoordinator;
use IAIso\Coordination\Snapshot;
use IAIso\Policy\SumAggregator;
use PHPUnit\Framework\TestCase;

final class CoordinationTest extends TestCase
{
    public function testAggregatesSum(): void
    {
        $c = SharedPressureCoordinator::builder()
            ->auditSink(new MemorySink())->build();
        $c->register('a');
        $c->register('b');
        $c->update('a', 0.3);
        $snap = $c->update('b', 0.5);
        self::assertEqualsWithDelta(0.8, $snap->aggregatePressure, 1e-9);
    }

    public function testEscalationCallbackFires(): void
    {
        $calls = 0;
        $c = SharedPressureCoordinator::builder()
            ->escalationThreshold(0.7)->releaseThreshold(0.95)
            ->notifyCooldownSeconds(0.0)
            ->onEscalation(function ($s) use (&$calls): void { $calls++; })
            ->build();
        $c->register('a');
        $c->update('a', 0.8);
        self::assertSame(1, $calls);
    }

    public function testRejectsBadPressure(): void
    {
        $c = SharedPressureCoordinator::builder()->build();
        $this->expectException(CoordinatorException::class);
        $c->update('a', 1.5);
    }

    public function testLuaScriptUnchangedFromSpec(): void
    {
        self::assertStringContainsString('pressures_key = KEYS[1]', RedisCoordinator::UPDATE_AND_FETCH_SCRIPT);
        self::assertStringContainsString('HGETALL', RedisCoordinator::UPDATE_AND_FETCH_SCRIPT);
        self::assertStringContainsString('EXPIRE', RedisCoordinator::UPDATE_AND_FETCH_SCRIPT);
    }

    public function testParseHGetAllFlatWorks(): void
    {
        $reply = ['a', '0.3', 'b', '0.5'];
        $out = RedisCoordinator::parseHGetAllFlat($reply);
        self::assertEqualsWithDelta(0.3, $out['a'], 1e-9);
        self::assertEqualsWithDelta(0.5, $out['b'], 1e-9);
    }

    public function testRedisCoordinatorWithMock(): void
    {
        $mock = new MockRedis();
        $c = RedisCoordinator::builder()
            ->redis($mock)->coordinatorId('test')
            ->escalationThreshold(0.7)->releaseThreshold(0.9)
            ->pressuresTtlSeconds(300)
            ->aggregator(new SumAggregator())
            ->auditSink(new MemorySink())
            ->build();
        $c->register('a');
        $c->register('b');
        $c->update('a', 0.4);
        $snap = $c->update('b', 0.3);
        self::assertEqualsWithDelta(0.7, $snap->aggregatePressure, 1e-9);
    }
}

/** Minimal in-memory Redis mock for tests. */
final class MockRedis implements RedisClient
{
    /** @var array<string,array<string,string>> */
    private array $hashes = [];

    public function eval(string $script, array $keys, array $args): mixed
    {
        $key = $keys[0];
        $this->hashes[$key] ??= [];
        if (str_contains($script, 'HSET') && str_contains($script, 'HGETALL')) {
            $this->hashes[$key][$args[0]] = $args[1];
            $flat = [];
            foreach ($this->hashes[$key] as $k => $v) {
                $flat[] = $k;
                $flat[] = $v;
            }
            return $flat;
        }
        if (str_contains($script, 'HDEL')) {
            unset($this->hashes[$key][$args[0]]);
            return 1;
        }
        return null;
    }

    public function hset(string $key, array $pairs): void
    {
        $this->hashes[$key] ??= [];
        foreach ($pairs as $p) {
            $this->hashes[$key][$p[0]] = $p[1];
        }
    }

    public function hkeys(string $key): array
    {
        return array_keys($this->hashes[$key] ?? []);
    }
}

File: IAIso-v5.0/core/iaiso-ruby/test/coordination_test.rb
# frozen_string_literal: true

require_relative "test_helper"
require "iaiso/audit"
require "iaiso/coordination"
require "iaiso/policy"

class CoordinationTest < Minitest::Test
  def test_aggregates_sum
    c = IAIso::Coordination::SharedPressureCoordinator.new(
      audit_sink: IAIso::Audit::MemorySink.new,
    )
    c.register("a")
    c.register("b")
    c.update("a", 0.3)
    snap = c.update("b", 0.5)
    assert_in_delta 0.8, snap.aggregate_pressure, 1e-9
  end

  def test_escalation_callback_fires
    calls = 0
    c = IAIso::Coordination::SharedPressureCoordinator.new(
      escalation_threshold: 0.7, release_threshold: 0.95,
      notify_cooldown_seconds: 0.0,
      on_escalation: ->(_s) { calls += 1 },
    )
    c.register("a")
    c.update("a", 0.8)
    assert_equal 1, calls
  end

  def test_rejects_bad_pressure
    c = IAIso::Coordination::SharedPressureCoordinator.new
    assert_raises(IAIso::Coordination::CoordinatorError) { c.update("a", 1.5) }
  end

  def test_lua_script_unchanged_from_spec
    s = IAIso::Coordination::RedisCoordinator::UPDATE_AND_FETCH_SCRIPT
    assert_includes s, "pressures_key = KEYS[1]"
    assert_includes s, "HGETALL"
    assert_includes s, "EXPIRE"
  end

  def test_parse_hgetall_flat
    reply = ["a", "0.3", "b", "0.5"]
    out = IAIso::Coordination::RedisCoordinator.parse_hgetall_flat(reply)
    assert_in_delta 0.3, out["a"], 1e-9
    assert_in_delta 0.5, out["b"], 1e-9
  end

  def test_redis_coordinator_with_mock
    mock = MockRedis.new
    c = IAIso::Coordination::RedisCoordinator.new(
      redis: mock, coordinator_id: "test",
      escalation_threshold: 0.7, release_threshold: 0.9,
      pressures_ttl_seconds: 300,
      audit_sink: IAIso::Audit::MemorySink.new,
    )
    c.register("a")
    c.register("b")
    c.update("a", 0.4)
    snap = c.update("b", 0.3)
    assert_in_delta 0.7, snap.aggregate_pressure, 1e-9
  end

  # Minimal in-memory mock for testing
  class MockRedis
    def initialize
      @hashes = {}
    end

    def eval(script, keys:, args:)
      key = keys[0]
      @hashes[key] ||= {}
      if script.include?("HSET") && script.include?("HGETALL")
        @hashes[key][args[0]] = args[1]
        flat = []
        @hashes[key].each { |k, v| flat << k << v }
        return flat
      end
      if script.include?("HDEL")
        @hashes[key].delete(args[0])
        return 1
      end
      nil
    end

    def hset(key, pairs)
      @hashes[key] ||= {}
      pairs.each { |p| @hashes[key][p[0]] = p[1] }
    end

    def hkeys(key)
      (@hashes[key] || {}).keys
    end
  end
end
